### PR TITLE
Azuretableacl

### PR DESCRIPTION
**Pull request by kpfaulkner**
Set/Gets ACLs for Storage Tables

Related to issue https://github.com/Azure/azure-sdk-for-go/issues/467

### CODE REVIEW

File: storage/table.go [hunk 1]
```diff
@@ -127,3 +143,109 @@ func (c *TableServiceClient) DeleteTable(table AzureTable) error {
...
+
+		if resp.statusCode != http.StatusNoContent {
+			err = errors.New("Unable to set permissions")
+			return err
+		}
+	}
+	return nil
+}
+
+// GetTablePermissions gets the table ACL permissions
```
Comment (mcardosos) [inline storage/table.go]: Here it is the other documentation link
https://docs.microsoft.com/en-us/rest/api/storageservices/fileservices/get-table-acl

File: storage/table_test.go [hunk 2]
```diff
@@ -285,3 +286,65 @@ func randTable() string {
 	}
 	return string(bytes)
 }
+
+func (s *StorageBlobSuite) createTablePermissions(ID string, canRead bool, canAppend bool, canUpdate bool,
+	canDelete bool, startTime time.Time, expiryTime time.Time) TableAccessPolicy {
+	policy := TableAccessPolicy{}
```
Comment (mcardosos) [inline storage/table_test.go]: You can instead use
```go
return TableAccessPolicy{
    ID: ID,
    StartTime: startTime,
    ExpiryTime: expiryTime,
    CanRead: canRead,
    CanAppend: canAppend,
    CanUpdate: canUpdate,
    CanDelete: canDelete,
}
```

File: storage/client.go [hunk 3]
```diff
@@ -308,6 +308,15 @@ func (c Client) buildCanonicalizedResourceTable(uri string) (string, error) {
 		cr += u.EscapedPath()
 	}
 
+	params, err := url.ParseQuery(u.RawQuery)
+
+	// search for "comp" parameter, if exists then add it to canonicalizedresource
+	for key := range params {
```
Comment (kpfaulkner) [inline storage/client.go]: I'm unclear what you mean by saying use Add? For which part?

As for the "comp" part, that is required when generating the hmac. Other parameters (if any exist) cannot be included, but "comp" is required.  Was a long hard struggle figuring that out.

Comment (mcardosos) [inline storage/client.go]: Interesting. @tamram could you add a bit more detail, please?

Comment (mcardosos) [inline storage/client.go]: From [Authentication for the Azure Storage Services](https://docs.microsoft.com/en-us/rest/api/storageservices/fileservices/authentication-for-the-azure-storage-services) (the "Constructing the Canonicalized Resource String" section, I can guess it has to do with that...

> Any portion of the CanonicalizedResource string that is derived from the resource's URI should be encoded exactly as it is in the URI.

Still asking the Storage team to understand this better.

Comment (kpfaulkner) [inline storage/client.go]: The only way I figured out what to do was to disassemble the dotnet binaries (then check github... duh)....    see https://github.com/Azure/azure-storage-net/blob/68c3ee55a3a6f62a0159cea58005d3fe027312a8/Lib/Common/Core/Util/AuthenticationUtility.cs#L313

After I did the same, everything sharedkeylite started working for me.

File: storage/table.go [hunk 4]
```diff
@@ -127,3 +143,109 @@ func (c *TableServiceClient) DeleteTable(table AzureTable) error {
...
+
+	var resp *odataResponse
+
+	// temp logging to figure out why dying in travis
+	log.Printf("access policy %s", accessPolicyXML)
+
+	if accessPolicyXML != "" {
+		headers["Content-Length"] = strconv.Itoa(len(accessPolicyXML))
+		resp, err = c.client.execTable("PUT", uri, headers, strings.NewReader(accessPolicyXML))
+	} else {
```
Comment (kpfaulkner) [inline storage/table.go]: you're absolutely right. I think I was in "container mode" when I wrote that (those can have items other than the policy XML). Removing.

File: storage/table.go [hunk 5]
```diff
@@ -127,3 +143,109 @@ func (c *TableServiceClient) DeleteTable(table AzureTable) error {
 	}
 	return nil
 }
+
+// SetTablePermissions sets up table ACL permissinos.
```
Comment (kpfaulkner) [inline storage/table.go]: done

File: storage/table.go [hunk 6]
```diff
@@ -127,3 +143,109 @@ func (c *TableServiceClient) DeleteTable(table AzureTable) error {
...
+func (c *TableServiceClient) SetTablePermissions(table AzureTable, accessPolicy TableAccessPolicy, timeout int) (err error) {
```
Comment (kpfaulkner) [inline storage/table.go]: done

File: storage/table.go [hunk 7]
```diff
@@ -127,3 +143,109 @@ func (c *TableServiceClient) DeleteTable(table AzureTable) error {
...
+	if resp != nil {
+		defer func() {
+			closeErr := resp.body.Close()
+			if closeErr != nil && err == nil {
+				err = closeErr
+			}
+		}()
+
+		if resp.statusCode != http.StatusNoContent {
+			err = errors.New("Unable to set permissions")
```
Comment (kpfaulkner) [inline storage/table.go]: done

File: storage/table.go (hunk 1 above)
Comment (kpfaulkner) [inline storage/table.go]: done

File: storage/common.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
+package storage
```
Comment (kpfaulkner) [inline storage/common.go]: will leave until @jhendrixMSFT gives comment

File: storage/table_test.go (hunk 2 above)
Comment (kpfaulkner) [inline storage/table_test.go]: done

**Comment by kpfaulkner** [comment]
@mcardosos Can you explain the merge errors I'm now seeing?

**Comment by marstr** [comment]
Howdy, @kpfaulkner. I just ran into the same problem this evening. It seems that  "gas" stopped ignoring a handful of example files that contain faux passwords, creating what it sees as a security issue. Tomorrow morning I'm going to work on mitigating this issue. I'll keep you posted.

**Comment by marstr** [comment]
Following up, I opened a new issue [(#485)](https://github.com/Azure/azure-sdk-for-go/issues/485) to track these integration problems.

**Comment by marstr** [comment]
Alright, as mentioned in Issue (#485) I've mitigated the problem that was causing build fails by disabling gas. If you rebase your changes on top of the latest master, that problem should go away. I'll re-enable gas once I can get it to start behaving again.

**Comment by kpfaulkner** [comment]
Cheers. Merged in master again, all good.

File: storage/client.go (hunk 3 above)
Comment (mcardosos) [inline storage/client.go]: Could you also add a test for this inside `Test_buildCanonicalizedResourceTable` in `client_test.go`, please?

Comment (kpfaulkner) [inline storage/client.go]: done

File: storage/common.go (hunk 8 above)
Comment (mcardosos) [inline storage/common.go]: Maybe a `storagepolicy.go` file?

Comment (kpfaulkner) [inline storage/common.go]: I'm unsure if the AccessPolicyXML is used in other places other than Tables and Blob (I'll have a look at Queues later once I'm back online). Agree that the common.go naming leaves a bit to be desired. I like the idea of storagepolicy.go as per @mcardosos 

Even if it does turn up queues use it, the name of storagepolicy.go is still valid and descriptive.

Any objections if I rename and push?

**Comment by mcardosos** [comment]
Merged!
@kpfaulkner Thanks for your always enthusiastic contributions!
:D

**Pull request merged**